### DISCUSSION TITLE
Values don't properly represent input

### DISCUSSION THREAD

**Discussion by JoshLudahl** [patrykandpatrick/vico#509] [Questions] [answered]
First off, let me thank you for your work, it is quite powerful! Thank you!

I have an application for weather. I'm implementing the Vico chart using the View system. I'm getting strange behavior when I input the data. 

So, I have a dataset of 48 dates that represent 48 hours of weather data. I'm pairing the hour to a precipitation amount. The time is in Unix format as an Int, the precipitation amount is a Double.

I am seeing weird behavior. When I shrink the dataset to something like 4 items, then I no longer see the values on the left axis. The hours are displayed properly though. When I use the full list of 48 items, I see the values on the left axis, but the bottom axis has way more items than my list. Like it almost endlessly scrolls, and the keys are incorrectly labeled. 

I have attached two screenshots so you can see what I am seeing.

![Screenshot_20231226_215233](https://github.com/patrykandpatrick/vico/assets/13421438/a9c1aa8d-1f61-4b8b-bb06-f10372fc4438)
![Screenshot_20231226_215209](https://github.com/patrykandpatrick/vico/assets/13421438/ab9c403f-8d4d-4551-a39a-97a14332c05a)

I also trimmed my code into one function, so you can see it all at once.

Code:

Fragment:

```
private fun buildCharts(result: WeatherDataSource) {
        
        // Create the list from the weather data source for the hourly data
        // Associate the date object with the rain precipitation
        val list = result.hourly.toList().subList(0, 4)
            .associate { it.dt to (it.rain?.h ?: 0f) }

        // Associate the keys to the date
        val xValuesToDates = list.keys
            .associateBy { it.toFloat() }

        // Create a chart entry model of the data, mapping the  values
        val chartEntryValues = xValuesToDates.keys.zip(list.values, ::entryOf)

        // Create a axis formatter
        val horizontalAxisValueFormatter =
            AxisValueFormatter<AxisPosition.Horizontal.Bottom> { value, _ ->

                // Convert the Int value to a date object
                var hour = Instant.ofEpochSecond(
                    (xValuesToDates[value]?.toLong() ?: value.toLong()) + result
                        .timezoneOffset
                )
                    .atZone(ZoneId.of("UTC"))
                    .hour

                if (hour > 12) hour -= 12
                if (hour == 0) hour = 12

                hour.toString()
            }

        (binding.chartView.bottomAxis as HorizontalAxis<AxisPosition.Horizontal.Bottom>)
            .valueFormatter = horizontalAxisValueFormatter

        val chartEntryModelProducer = ChartEntryModelProducer()
        val chartEntryModel = entryModelOf(chartEntryValues)
        chartEntryModelProducer.setEntries(chartEntryModel.entries)
        binding.chartView.setModel(chartEntryModel)
    }
```

Layout:
```
<com.patrykandpatrick.vico.views.chart.ChartView
                            android:id="@+id/chart_view"
                            android:layout_width="wrap_content"
                            android:layout_height="wrap_content"
                            android:layout_marginTop="8dp"
                            android:padding="16dp"
                            android:background="@drawable/top_rounded_rectangle"
                            app:chart="column"
                            app:layout_constraintEnd_toEndOf="parent"
                            app:layout_constraintStart_toStartOf="parent"
                            app:layout_constraintTop_toBottomOf="@+id/hourly_label"
                            app:columnChartStyle="@style/chart_style"
                            app:showBottomAxis="true"
                            app:showStartAxis="true" />
```

Any idea what is causing this? I have tried a few different transformations of the data, but nothing has seemed to work as expected. Maybe I am missing something here?

**Comment by patrickmichalik** [comment]
Hello! Thanks for the kind words. Could you please share the content of `chartEntryValues`? This should suffice:

```kt
Log.d("buildCharts", chartEntryValues)
```

**Reply by JoshLudahl** [reply]
Hello, thanks for the reply!

Here is the output:

```
2023-12-27 08:18:09.587  9605-9605  buildCharts             com.softklass.warbler                D  [FloatEntry(x=1.7036928E9, y=0.14), FloatEntry(x=1.70369638E9, y=1.0), FloatEntry(x=1.70369997E9, y=0.0), FloatEntry(x=1.70370355E9, y=0.0), FloatEntry(x=1.70370714E9, y=0.2), FloatEntry(x=1.70371085E9, y=1.89), FloatEntry(x=1.70371443E9, y=1.43), FloatEntry(x=1.70371802E9, y=0.0), FloatEntry(x=1.7037216E9, y=0.0), FloatEntry(x=1.70372518E9, y=0.0), FloatEntry(x=1.70372877E9, y=0.0), FloatEntry(x=1.70373235E9, y=0.0), FloatEntry(x=1.70373606E9, y=0.0), FloatEntry(x=1.70373965E9, y=0.0), FloatEntry(x=1.70374323E9, y=0.0), FloatEntry(x=1.70374682E9, y=0.0), FloatEntry(x=1.7037504E9, y=0.0), FloatEntry(x=1.70375398E9, y=0.0), FloatEntry(x=1.70375757E9, y=0.0), FloatEntry(x=1.70376115E9, y=0.0), FloatEntry(x=1.70376474E9, y=0.0), FloatEntry(x=1.70376845E9, y=0.0), FloatEntry(x=1.70377203E9, y=0.0), FloatEntry(x=1.70377562E9, y=0.0), FloatEntry(x=1.7037792E9, y=0.0), FloatEntry(x=1.70378278E9, y=0.0), FloatEntry(x=1.70378637E9, y=0.0), FloatEntry(x=1.70378995E9, y=0.0), FloatEntry(x=1.70379366E9, y=0.0), FloatEntry(x=1.70379725E9, y=0.0), FloatEntry(x=1.70380083E9, y=0.0), FloatEntry(x=1.70380442E9, y=0.0), FloatEntry(x=1.703808E9, y=0.0), FloatEntry(x=1.70381158E9, y=0.0), FloatEntry(x=1.70381517E9, y=0.0), FloatEntry(x=1.70381875E9, y=0.0), FloatEntry(x=1.70382234E9, y=0.0), FloatEntry(x=1.70382605E9, y=0.0), FloatEntry(x=1.70382963E9, y=0.0), FloatEntry(x=1.70383322E9, y=0.0), FloatEntry(x=1.7038368E9, y=0.0), FloatEntry(x=1.70384038E9, y=0.0), FloatEntry(x=1.70384397E9, y=0.0), FloatEntry(x=1.70384755E9, y=0.0), FloatEntry(x=1.70385126E9, y=0.0), FloatEntry(x=1.70385485E9, y=0.0), FloatEntry(x=1.70385843E9, y=0.0), FloatEntry(x=1.70386202E9, y=0.0)]
```
As you can see, the list size is the 48 items I expect. But when the chart shows, the list almost endlessly scrolls the bottom axis, which is far beyond the 48 items. 

Let me know if anything else would be useful. Thank you.

**Reply by JoshLudahl** [reply]
I updated the library to the latest (alpha.5), I am still seeing the issue of having way more values than input.

```
  private fun setUpNewChart(result: WeatherDataSource) {
        val data =
            result.hourly.toList().associate {
                it.dt.toLong() to (it.rain?.h ?: 0.0)
            }

        val xValuesToDates = data.keys.associateBy { it.toFloat() }

        val model =
            CartesianChartModel(
                ColumnCartesianLayerModel
                    .build { series(x = xValuesToDates.keys, y = data.values) },
            )

        val bottomAxisValueFormatter =
            AxisValueFormatter<AxisPosition.Horizontal.Bottom> { x, _, _ ->
                // Convert the Int value to a date object
                var hour =
                    Instant.ofEpochSecond(
                        (xValuesToDates[x]?.toLong() ?: x.toLong()) +
                            result
                                .timezoneOffset,
                    )
                        .atZone(ZoneId.of("UTC"))
                        .hour

                if (hour > 12) hour -= 12
                if (hour == 0) hour = 12

                hour.toString()
            }

        
        binding.chartView.setModel(model)

        (binding.chartView.chart?.bottomAxis as HorizontalAxis<AxisPosition.Horizontal.Bottom>)
            .valueFormatter = bottomAxisValueFormatter
    }
```

I have updated the code block I am using now.

Attached is the screenshot of the repetitive values. 
![Screenshot_20231227_152738](https://github.com/patrykandpatrick/vico/assets/13421438/f2b6ab3f-e09a-4c39-98b2-4d5578930fb0)

**Reply by patrickmichalik** [reply]
Thanks, @JoshLudahl! I posted a reply yesterday but deleted it after realizing it wasn’t entirely accurate. Sorry about it. I’ve taken a closer look, and now I know what’s going on.

It’s great that you’ve updated to Vico 2.0.0 Alpha 5—it resolves a few issues that were playing into this. There should now be no more problems on the Vico side. The reason why you’re getting so many labels is that the _x_ values aren’t equally spaced. The numbers here are quite large, so let me give you a simpler example. Let’s say you have these _x_ values: [1, 3, 17, 19]. It would be mathematically inaccurate for these to be positioned at regular intervals. The difference between 17 and 2 is greater than that between 2 and 1, for example. Vico calculates the GCD of the differences between neighboring _x_ values and uses that as the _x_ step. You thus get labels for these _x_ values: [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]. It’s a similar story in your chart. There isn’t a common difference between neighboring timestamps. The GCD of the differences is found, and it’s relatively small. The labels you see along the _x_-axis are for different values, but your `AxisValueFormatter` appears to retain only the provided timestamp’s hour, so the same string would be produced for 3:00 p.m. and 3:01 p.m., for instance.

To achieve the desired result, you’ll have to change how the _x_ values are determined. If the columns are to be spaced out evenly, the same should apply to the _x_ values. I’ll prepare an example for you as soon as I get home.

**Reply by patrickmichalik** [reply, marked as answer]
Please try the following, @JoshLudahl:

```kt
private fun setUpNewChart(result: WeatherDataSource) {
    val rain = result.hourly.map { it.rain?.h ?: 0.0 }
    val hours =
        result.hourly.map {
            val hour = Instant.ofEpochSecond(it.dt.toLong() + result.timezoneOffset).atZone(ZoneId.of("UTC")).hour % 12
            (if (hour == 0) 12 else hour).toString()
        }
    val model = CartesianChartModel(ColumnCartesianLayerModel.build { series(rain) })
    val bottomAxisValueFormatter = AxisValueFormatter<AxisPosition.Horizontal.Bottom> { x, _, _ -> hours[x.toInt()] }
    binding.chartView.setModel(model)
    (binding.chartView.chart?.bottomAxis as HorizontalAxis<AxisPosition.Horizontal.Bottom>).valueFormatter =
        bottomAxisValueFormatter
}
```

I used some sample data (1 a.m. through 10 a.m.) and got this:

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/dd784996-9e35-4263-8c39-74823c92bdec" width="720" />

**Comment by JoshLudahl** [comment]
Perfect, thank you. I will give this a try. I was overthinking the process
and simplifying the input and calculating the formatting makes more sense.
Thank you again!

**Reply by patrickmichalik** [reply]
You’re most welcome!
